openraft/src/engine/log_id_list.rs

```diff
@@ -0,0 +1,105 @@
+use crate::raft_types::RaftLogId;
+use crate::LogId;
+use crate::NodeId;
+
+/// Efficient storage for log ids.
+///
+/// It stores only the ids of log that have a new leader_id. And the `last_log_id` at the end.
+/// I.e., the oldest log id belonging to every leader.
+///
+/// If it is not empty, the first one is `last_purged_log_id` and the last one is `last_log_id`.
+/// The last one may have the same leader id as the second last one.
+#[derive(Default, Debug, Clone)]
+#[derive(PartialEq, Eq)]
+pub struct LogIdList<NID: NodeId> {
```

Comment:
seems `LogIdList` is not a simple log id list, it does not simply save log ids, but key log id, so IMHO `LogIdList` is not accuracy, may be `KeyLogIdList` is fine.

Reply:
and if rename `LogIdList`, may be the file name will be renamed too.

Reply:
I'd like a user not to need to understand what it is inside.

Its user should not have to be aware that actually not all the ids are stored.

Reply:
yes, but the source reader may be confused by the name

Reply:
> yes, but the source reader may be confused by the name

Right...Actually I named it `KeyLogIdList` and renamed it to `LogIdList`.

I'd still prefer the shorter one, to purely express what a struct does instead of explaining how it does in a name.

For the same reason, `LeaderState` is not a good name in this criterion because it does not explain what it does, IHMO, although a name such as `TheStateAfterVoteGrantedByQuorum` would be more obvious. :(